follow_sponsorship/model/contracts.py

```diff
@@ -0,0 +1,277 @@
+
+        for contract in self.browse(cr, uid, contract_ids, context):
+            if contract.end_date:
+                end_date = datetime.strptime(
+                    contract.end_date, DF).date()
+                if end_date + timedelta(days=15) < date.today():
+                    wf_service = netsvc.LocalService('workflow')
+                    logger.info(
+                        "Contract " + str(
+                            contract.id) + " sub waiting time expired")
+                    wf_service.trg_validate(
+                        uid, 'recurring.contract',
+                        contract.id, 'no_sub', cr)
+        return True
+
+    def check_sub_duration(self, cr, uid, context=None):
+        contract_ids = self.search(
+            cr, uid, [('sds_state', '=', 'sub')], context=context)
+
+        for contract in self.browse(cr, uid, contract_ids, context):
+            if contract.date_sub:
+                date_sub = datetime.strptime(
+                    contract.date_sub, DF).date()
+                if date_sub + timedelta(days=15) < date.today():
+                    wf_service = netsvc.LocalService('workflow')
+                    logger.info(
+                        "Contract " + str(
+                            contract.id) + " sub waiting time expired")
+                    wf_service.trg_validate(
+                        uid, 'recurring.contract',
+                        contract.id,
+                        'sub_accept', cr)
+        return True
+
+    def contract_cancelled(self, cr, uid, ids, context=None):
+        res = super(recurring_contract, self).contract_cancelled(
+            cr, uid, ids, context)
+        for contract in self.browse(cr, uid, ids, context):
+            if contract.parent_id:
+                wf_service = netsvc.LocalService('workflow')
+                logger.info("Contract " + str(contract.id) + " sub rejected")
+                wf_service.trg_validate(
+                    uid, 'recurring.contract',
+                    contract.parent_id.id,
+                    'sub_reject', cr)
+        return res
+
+    def on_change_partner_id(self, cr, uid, ids, partner_id, context=None):
+        res = super(recurring_contract, self).on_change_partner_id(
+            cr, uid, ids, partner_id, context)
+        origin_ids = self.pool.get('recurring.contract.origin').search(
+            cr, uid,
+            [('name', '=', 'SUB Sponsorship')],
+            context=context)
+        parent_id = self.define_parent_id(cr, uid, partner_id, context)
+        origin_id = origin_ids[0] if parent_id else False
+
+        res['value'].update({
+            'parent_id': parent_id,
+            'origin_id': origin_id
+        })
+        return res
+
+    def define_parent_id(self, cr, uid, partner_id, context=None):
+        same_partner_contracts_ids = self.search(
+            cr, uid,
+            [('partner_id', '=', partner_id),
+             ('state', '=', 'terminated')],

```

Comment:
Also include contracts in 'cancelled' state as a depart could happen before the sponsor has paid.